### PR TITLE
Rules API: Allow filtering by rule name

### PR DESCRIPTION
**Pull request by gotjosh**
Introduces support for a new query parameter in the `/rules` API endpoint that allows filtering by rule names.

If all the rules of a group are filtered, we skip the group entirely.

### CODE REVIEW

File: web/api/v1/api.go [hunk 1]
```diff
@@ -1296,6 +1296,16 @@ func (api *API) rules(r *http.Request) apiFuncResult {
 	res := &RuleDiscovery{RuleGroups: make([]*RuleGroup, len(ruleGroups))}
 	typ := strings.ToLower(r.URL.Query().Get("type"))
 
+	// Parse the rule names into a comma separated list of rule names, then create a set.
+	rulesQuery := r.URL.Query().Get("rules")
+	ruleNamesSet := map[string]struct{}{}
+	if rulesQuery != "" {
+		names := strings.Split(rulesQuery, ",")
```
Comment (gotjosh) [inline web/api/v1/api.go:1303]: `Split` returns a 1 element slice with `"` so it can't be inlined.

File: web/api/v1/api.go [hunk 2]
```diff
@@ -1313,14 +1323,20 @@ func (api *API) rules(r *http.Request) apiFuncResult {
 			EvaluationTime: grp.GetEvaluationTime().Seconds(),
 			LastEvaluation: grp.GetLastEvaluation(),
 		}
-		for _, r := range grp.Rules() {
+		for _, rr := range grp.Rules() {
```
Comment (gotjosh) [inline web/api/v1/api.go:1326]: Drive-by fix because `r` was shadoed by `r *http.Request`

File: docs/querying/api.md [hunk 3]
```diff
@@ -673,7 +673,9 @@ GET /api/v1/rules
 ```
 
 URL query parameters:
+
 - `type=alert|record`: return only the alerting rules (e.g. `type=alert`) or the recording rules (e.g. `type=record`). When the parameter is absent or empty, no filtering is done.
+- `rules=alertName,RuleName`: return only the alerting and recording rules with the specified names. If we've filtered out all the rules of a group, the group is not returned. When the parameter is absent or empty, no filtering is done.
```
Comment (juliusv) [inline docs/querying/api.md:678]: Rule names may contain commas, so we would have to introduce escaping to make this fully correct. But I would propose just allowing the parameter to be repeated, like we do for the `match[]` parameter in various API endpoints (see https://prometheus.io/docs/prometheus/latest/querying/api/#finding-series-by-label-matchers for example). That's both an established pattern and keeps things simpler.

The question is then whether `rule[]` would be a good name for the repeated filter parameter. Maybe `rule_name[]` to make it more explicit what we're filtering on, especially in case there's other future filters for this endpoint (e.g. I guess it would make sense to also filter by rule group name at the same time to fully scope your query and not accidentally select matching rule names from an unrelated group?).

Regarding naming I also noticed that we accidentally(?) started using both camelCase (e.g. `scrapePool` in the `/api/v1/targets` endpoint) and snake_case (e.g. `match_target` in `/api/v1/targets/metadata`) for HTTP parameters. But snake case was there first and is also more consistent with the HTTP path naming that we use for our endpoints, so I would prefer it for future parameter names I think.

Comment (juliusv) [reply docs/querying/api.md:678]: Good point about the file name! Yeah, that sounds great to me :)

**Comment by juliusv** [comment]
@gotjosh Looks good now, just the docs need updating still.

File: docs/querying/api.md [hunk 4]
```diff
@@ -675,9 +675,9 @@ GET /api/v1/rules
 URL query parameters:
 
 - `type=alert|record`: return only the alerting rules (e.g. `type=alert`) or the recording rules (e.g. `type=record`). When the parameter is absent or empty, no filtering is done.
-- `rule_name[]=<rule name>`: return the groups and rules of the specified alerting and recording rules names, the parameter supports repetition. If we've filtered out all the rules of a group, the group is not returned. When the parameter is absent or empty, no filtering is done.
-- `rule_group[]=<group name>`: return the group and rules of the specified rule group names, the parameter supports repetitions. When the parameter is absent or empty, no filtering is done.
-- `file[]=<filepath>`: return the group and rules of specified filepath for rule groups, the parameter supports repetition. When the parameter is absent or empty, no filtering is done.
+- `rule_name[]=<rule name>`: return the groups and its rules of the specified alerting and recording rules names, the parameter supports repetition. If we've filtered out all the rules of a group, the group is not returned. When the parameter is absent or empty, no filtering is done.
```
Comment (juliusv) [inline docs/querying/api.md:678]: For the placeholder I would just use `<string>` everyhwere, because otherwise we usually define it at the top of the page as a new placeholder type.

Hmm, and I think not all of the plural / singular grammar matches up in the description ("groups and its" and "rules names")? How about something simpler like:

```suggestion
- `rule_name[]=<string>`: only return rules with the given rule name. If the parameter is repeated, rules with any of the provided names are returned. If we've filtered out all the rules of a group, the group is not returned. When the parameter is absent or empty, no filtering is done.
```

And similar for the others.

Comment (gotjosh) [reply docs/querying/api.md]: I have done a bit more wordsmithing:

- removed the filtering comment from files and groups - it only applies to rules.
- included your suggested wording for the first sentence
- included the entity name in the second sentence

My only nit is that the first sentence is not entirely accurate as suggested - what we return in all cases is rule groups (because it's what the API returns) so saying something like "only return rules" is not entirely accurate, but perhaps this is OK? As in reality, what people come to look at here is rules.

File: docs/querying/api.md [hunk 5]
```diff
@@ -673,7 +673,11 @@ GET /api/v1/rules
 ```
 
 URL query parameters:
+
 - `type=alert|record`: return only the alerting rules (e.g. `type=alert`) or the recording rules (e.g. `type=record`). When the parameter is absent or empty, no filtering is done.
+- `rule_name[]=<string>`: only return rules with the given rule name. If the parameter is repeated, rules with any of provided names are returned. If we've filtered out all the rules of a group, the group is not returned. When the parameter is absent or empty, no filtering is done.
```
Comment (juliusv) [inline docs/querying/api.md:678]: Here and below:

```suggestion
- `rule_name[]=<string>`: only return rules with the given rule name. If the parameter is repeated, rules with any of the provided names are returned. If we've filtered out all the rules of a group, the group is not returned. When the parameter is absent or empty, no filtering is done.
```

:+1: Feel free to merge after fixing this nit :)

**Comment by gotjosh** [comment]
Alright, did some more manual testing and found out filtered rule groups where showing up as `null` on the response. Tried many different combinations and they all work as expected using these files:

<details>
<summary>
See files
</summary>


`first.rules`
```yaml
groups:
  - name: test
    rules:
      - record: job:http_requests:rate5m
        expr: sum by (job)(rate(http_requests_total[5m]))
  - name: test2
    rules:
      - record: job:http_requests:rate5m
        expr: sum by (job)(rate(http_requests_total[5m]))
  - name: test3
    rules:
      - record: job:http_requests:rate5m
        expr: sum by (job)(rate(http_requests_total[5m]))
  - name: none
    rules:
      - record: job:http_requests:rate5m
        expr: sum by (job)(rate(http_requests_total[5m]))
```

`second.rules`
```yaml
groups:
  - name: test
    rules:
      - record: job:http_requests:rate5m
        expr: sum by (job)(rate(http_requests_total[5m]))
  - name: test2
    rules:
      - record: job:http_requests:rate5m
        expr: sum by (job)(rate(http_requests_total[5m]))
  - name: test3
    rules:
      - record: job:http_requests:rate5m
        expr: sum by (job)(rate(http_requests_total[5m]))
      - record: testrule
        expr: sum by (job)(rate(http_requests_total[5m]))
      - record: onemore
        expr: sum by (job)(rate(http_requests_total[5m]))
  - name: different
    rules:
      - record: onemore
        expr: sum by (job)(rate(http_requests_total[5m]))
      - record: none
        expr: sum by (job)(rate(http_requests_total[5m]))
```

then a config of:

```yaml
# my global config
global:
  scrape_interval: 15s
  evaluation_interval: 30s
  # scrape_timeout is set to the global default (10s).

  external_labels:
    monitor: codelab
    foo: bar

rule_files:
  - "first.rules"
  - "second.rules"

```

</details>


I adjusted the test, and we're all good. Want to take one last quick look @juliusv ?

**Comment by juliusv** [comment]
Ah, good catch! Looks good now :) :+1:
